In [1]:
import pandas as pd
import numpy as np

In [2]:
df =pd.read_csv(r'Dataset.csv')
df

Restaurant ID           Restaurant Name  Country Code              City  \
0           6317637          Le Petit Souffle           162       Makati City   
1           6304287          Izakaya Kikufuji           162       Makati City   
2           6300002    Heat - Edsa Shangri-La           162  Mandaluyong City   
3           6318506                      Ooma           162  Mandaluyong City   
4           6314302               Sambo Kojin           162  Mandaluyong City   
...             ...                       ...           ...               ...   
9546        5915730               Naml۱ Gurme           208         ��stanbul   
9547        5908749              Ceviz A��ac۱           208         ��stanbul   
9548        5915807                     Huqqa           208         ��stanbul   
9549        5916112               A���k Kahve           208         ��stanbul   
9550        5927402  Walter's Coffee Roastery           208         ��stanbul   

                                                Address  \
0     Third Floor, Century City Mall, Kalayaan Avenu...   
1     Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2     Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3     Third Floor, Mega Fashion Hall, SM Megamall, O...   
4     Third Floor, Mega Atrium, SM Megamall, Ortigas...   
...                                                 ...   
9546  Kemanke�� Karamustafa Pa��a Mahallesi, R۱ht۱m ...   
9547  Ko��uyolu Mahallesi, Muhittin ��st�_nda�� Cadd...   
9548  Kuru�_e��me Mahallesi, Muallim Naci Caddesi, N...   
9549  Kuru�_e��me Mahallesi, Muallim Naci Caddesi, N...   
9550  Cafea��a Mahallesi, Bademalt۱ Sokak, No 21/B, ...   

                                        Locality  \
0      Century City Mall, Poblacion, Makati City   
1     Little Tokyo, Legaspi Village, Makati City   
2     Edsa Shangri-La, Ortigas, Mandaluyong City   
3         SM Megamall, Ortigas, Mandaluyong City   
4         SM Megamall, Ortigas, Mandaluyong City   
...                                          ...   
9546                                    Karak�_y   
9547                                   Ko��uyolu   
9548                                 Kuru�_e��me   
9549                                 Kuru�_e��me   
9550                                        Moda   

                                       Locality Verbose   Longitude  \
0     Century City Mall, Poblacion, Makati City, Mak...  121.027535   
1     Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101   
2     Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831   
3     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475   
4     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508   
...                                                 ...         ...   
9546                                Karak�_y, ��stanbul   28.977392   
9547                               Ko��uyolu, ��stanbul   29.041297   
9548                             Kuru�_e��me, ��stanbul   29.034640   
9549                             Kuru�_e��me, ��stanbul   29.036019   
9550                                    Moda, ��stanbul   29.026016   

       Latitude                          Cuisines  ...          Currency  \
0     14.565443        French, Japanese, Desserts  ...  Botswana Pula(P)   
1     14.553708                          Japanese  ...  Botswana Pula(P)   
2     14.581404  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)   
3     14.585318                   Japanese, Sushi  ...  Botswana Pula(P)   
4     14.584450                  Japanese, Korean  ...  Botswana Pula(P)   
...         ...                               ...  ...               ...   
9546  41.022793                           Turkish  ...  Turkish Lira(TL)   
9547  41.009847   World Cuisine, Patisserie, Cafe  ...  Turkish Lira(TL)   
9548  41.055817            Italian, World Cuisine  ...  Turkish Lira(TL)   
9549  41.057979                   Restaurant Cafe  ...  Turkish Lira(TL)   
9550  40.9

In [3]:
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

In [4]:
df.duplicated().sum()

0

In [5]:
import re
def remove_special_chars(text):
    if isinstance(text, str):
        return re.sub(r'[^\w\s]', '', text)
    else:
        return text

In [6]:
df_clean = df.applymapdf_clean = df.applymap(remove_special_chars)
df_clean(remove_special_chars)
df_clean

C:\Users\kumar\AppData\Local\Temp\ipykernel_22832\2139489135.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_clean = df.applymap(remove_special_chars)


Restaurant ID          Restaurant Name  Country Code              City  \
0           6317637         Le Petit Souffle           162       Makati City   
1           6304287         Izakaya Kikufuji           162       Makati City   
2           6300002     Heat  Edsa ShangriLa           162  Mandaluyong City   
3           6318506                     Ooma           162  Mandaluyong City   
4           6314302              Sambo Kojin           162  Mandaluyong City   
...             ...                      ...           ...               ...   
9546        5915730              Naml۱ Gurme           208           stanbul   
9547        5908749               Ceviz Aac۱           208           stanbul   
9548        5915807                    Huqqa           208           stanbul   
9549        5916112                 Ak Kahve           208           stanbul   
9550        5927402  Walters Coffee Roastery           208           stanbul   

                                                Address  \
0     Third Floor Century City Mall Kalayaan Avenue ...   
1     Little Tokyo 2277 Chino Roces Avenue Legaspi V...   
2     Edsa ShangriLa 1 Garden Way Ortigas Mandaluyon...   
3     Third Floor Mega Fashion Hall SM Megamall Orti...   
4     Third Floor Mega Atrium SM Megamall Ortigas Ma...   
...                                                 ...   
9546  Kemanke Karamustafa Paa Mahallesi R۱ht۱m Cadde...   
9547  Kouyolu Mahallesi Muhittin st_nda Caddesi No 8...   
9548  Kuru_eme Mahallesi Muallim Naci Caddesi No 56 ...   
9549  Kuru_eme Mahallesi Muallim Naci Caddesi No 64B...   
9550  Cafeaa Mahallesi Bademalt۱ Sokak No 21B Kad۱k_...   

                                      Locality  \
0      Century City Mall Poblacion Makati City   
1     Little Tokyo Legaspi Village Makati City   
2      Edsa ShangriLa Ortigas Mandaluyong City   
3         SM Megamall Ortigas Mandaluyong City   
4         SM Megamall Ortigas Mandaluyong City   
...                                        ...   
9546                                   Karak_y   
9547                                   Kouyolu   
9548                                  Kuru_eme   
9549                                  Kuru_eme   
9550                                      Moda   

                                       Locality Verbose   Longitude  \
0     Century City Mall Poblacion Makati City Makati...  121.027535   
1     Little Tokyo Legaspi Village Makati City Makat...  121.014101   
2     Edsa ShangriLa Ortigas Mandaluyong City Mandal...  121.056831   
3     SM Megamall Ortigas Mandaluyong City Mandaluyo...  121.056475   
4     SM Megamall Ortigas Mandaluyong City Mandaluyo...  121.057508   
...                                                 ...         ...   
9546                                    Karak_y stanbul   28.977392   
9547                                    Kouyolu stanbul   29.041297   
9548                                   Kuru_eme stanbul   29.034640   
9549                                   Kuru_eme stanbul   29.036019   
9550                                       Moda stanbul   29.026016   

       Latitude                       Cuisines  ...        Currency  \
0     14.565443       French Japanese Desserts  ...  Botswana PulaP   
1     14.553708                       Japanese  ...  Botswana PulaP   
2     14.581404  Seafood Asian Filipino Indian  ...  Botswana PulaP   
3     14.585318                 Japanese Sushi  ...  Botswana PulaP   
4     14.584450                Japanese Korean  ...  Botswana PulaP   
...         ...                            ...  ...             ...   
9546  41.022793                        Turkish  ...  Turkish LiraTL   
9547  41.009847  World Cuisine Patisserie Cafe  ...  Turkish LiraTL   
9548  41.055817          Italian World Cuisine  ...  Turkish LiraTL   
9549  41.057979                Restaurant Cafe  ...  Turkish LiraTL   
9550  40.984776                           Cafe  ...  Turkish LiraTL   

     Has Table booking Has On

In [7]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9551 entries, 0 to 9550
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Restaurant ID         9551 non-null   int64  
 1   Restaurant Name       9551 non-null   object 
 2   Country Code          9551 non-null   int64  
 3   City                  9551 non-null   object 
 4   Address               9551 non-null   object 
 5   Locality              9551 non-null   object 
 6   Locality Verbose      9551 non-null   object 
 7   Longitude             9551 non-null   float64
 8   Latitude              9551 non-null   float64
 9   Cuisines              9542 non-null   object 
 10  Average Cost for two  9551 non-null   int64  
 11  Currency              9551 non-null   object 
 12  Has Table booking     9551 non-null   object 
 13  Has Online delivery   9551 non-null   object 
 14  Is delivering now     9551 non-null   object 
 15  Switch to order menu 

In [8]:
features = ['Country Code', 'City', 'Cuisines', 'Average Cost for two', 'Has Table booking', 'Has Online delivery', 'Price range', 'Votes']

In [9]:
X = df[features]
y = df['Aggregate rating']

In [10]:
categorical_features = ['City', 'Cuisines', 'Has Table booking', 'Has Online delivery']

In [11]:
from sklearn.preprocessing import OneHotEncoder

In [12]:
one_hot = OneHotEncoder()

In [13]:
from sklearn.compose import ColumnTransformer

In [14]:
transformer = ColumnTransformer([("one_hot", one_hot, categorical_features)], remainder="passthrough")

In [15]:
X_transformed = transformer.fit_transform(X)

In [16]:
from sklearn.model_selection import train_test_split

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X_transformed, y, test_size=0.2, random_state=42)

In [19]:
from sklearn.linear_model import LinearRegression

In [20]:
model = LinearRegression()

In [21]:
model.fit(X_train, y_train)

LinearRegression()

In [22]:
y_pred = model.predict(X_test)
y_pred

array([1.34742238, 3.14761128, 1.86735685, ..., 1.80359861, 3.66711656,
       4.28103327])

In [23]:
from sklearn.metrics import mean_squared_error,r2_score

In [24]:
rmse = mean_squared_error(y_test, y_pred, squared=False)

In [25]:
from sklearn.metrics import mean_squared_error, r2_score

In [26]:
mse = mean_squared_error(y_test, y_pred)
r_squared = r2_score(y_test, y_pred)

In [27]:
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared: {r_squared}")

Mean Squared Error (MSE): 1.447478404496033
R-squared: 0.36405575890317154


In [28]:
coefficients = model.coef_

In [29]:
feature_names = transformer.named_transformers_['one_hot'].get_feature_names_out()

In [30]:
all_feature_names = feature_names.tolist() + ['Country Code', 'Average Cost for two', 'Price range', 'Votes']

In [31]:
coefficients_df = pd.DataFrame({
    'Feature': all_feature_names,
    'Coefficient': coefficients
})

In [32]:
coefficients_df['Absolute Coefficient'] = coefficients_df['Coefficient'].abs()
coefficients_df = coefficients_df.sort_values(by='Absolute Coefficient', ascending=False)
coefficients_df

Feature  Coefficient  \
89                                           City_Noida    -1.329376   
50                                         City_Gurgaon    -0.986002   
88                                       City_New Delhi    -0.875222   
43                                       City_Faridabad    -0.776907   
1447                              Cuisines_North Indian    -0.739084   
...                                                 ...          ...   
1062                              Cuisines_French, Kiwi     0.000000   
1069       Cuisines_Goan, Chinese, Indian, North Indian     0.000000   
1632  Cuisines_North Indian, Lebanese, European, Mex...     0.000000   
1072                             Cuisines_Goan, Seafood     0.000000   
529    Cuisines_Cafe, Continental, Italian, Street Food     0.000000   

      Absolute Coefficient  
89                1.329376  
50                0.986002  
88                0.875222  
43                0.776907  
1447              0.739084  
...                    ...  
1062              0.000000  
1069              0.000000  
1632              0.000000  
1072              0.000000  
529               0.000000  

[1975 rows x 3 columns]

Most Influential Features Affecting Restaurant Ratings:
City_Noida:

Coefficient: -1.329376
Absolute Coefficient: 1.329376
This suggests that being located in Noida has a significant negative impact on restaurant ratings.
City_Gurgaon:

Coefficient: -0.986002
Absolute Coefficient: 0.986002
Restaurants in Gurgaon also have a negative impact on ratings, but slightly less than Noida.
City_New Delhi:

Coefficient: -0.875222
Absolute Coefficient: 0.875222
Being located in New Delhi negatively affects ratings as well, but less so than Noida and Gurgaon.
City_Faridabad:

Coefficient: -0.776907
Absolute Coefficient: 0.776907
Faridabad also shows a negative influence on restaurant ratings.
Cuisines_North Indian:

Coefficient: -0.739084
Absolute Coefficient: 0.739084
Restaurants serving North Indian cuisine tend to have slightly lower ratings.